### ISSUE TITLE
NadamOptimizer does not work with sparse gradients

### ISSUE THREAD

**Issue by albertz** [tensorflow/tensorflow#15035]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: pip
- **TensorFlow version (use command below)**: v1.4.0-rc1-11-g130a514 1.4.0
- **Python version**: 3.6.3

Code:
```
import tensorflow as tf
from tensorflow.contrib.opt import NadamOptimizer

optimizer = NadamOptimizer(learning_rate=0.001)
# optimizer = tf.train.AdamOptimizer(learning_rate=0.001)  # this works
w = tf.get_variable("w", shape=(100, 10))
idxs = tf.placeholder(tf.int32, shape=(None,))
emb = tf.gather(w, idxs)
loss = tf.reduce_sum(emb ** 2)
minimize = optimizer.minimize(loss)

with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())
  sess.run(minimize, feed_dict={idxs: [1, 2, 3]})
```

This fails with:
```
...
  File "/u/zeyer/.local/lib/python3.6/site-packages/tensorflow/contrib/opt/python/training/nadam_optimizer.py", line 83, in _apply_sparse_shared
    m_bar = m_scaled_g_values + beta1_t * m_t
...

InvalidArgumentError (see above for traceback): Incompatible shapes: [3,10] vs. [100,10]
         [[Node: Adam/update_w/add = Add[T=DT_FLOAT, _class=["loc:@w"], _device="/job:localhost/replica:0/task:0/device:GPU:0"](Adam/update_w/mul_1, Adam/update_w/mul_3)]]
```

The bug is pretty obvious in `nadam_optimizer.py`. The fix would be to do it like in `adam.py`:
```
m_t = state_ops.assign(m, m * beta1_t, use_locking=self._use_locking)
with ops.control_dependencies([m_t]):
  m_t = scatter_add(m, indices, m_scaled_g_values)
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by albertz** [comment]
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: N/A

**Comment by reedwm** [comment]
@alextp can you fix this?

Also, @albertz, feel free to submit a PR to address this.

**Comment by alextp** [comment]
I'm not the maintainer of this contrib optimizer so I think this is contributions welcome.

**Comment by JingJunYin** [comment]
I created a PR for this, please feel free to review #15665

**Comment by sushreebarsa** [comment]
@albertz Please see the TensorFlow v1 to TensorFlow v2 [migration guide](https://www.tensorflow.org/guide/migrate).We see that you are using TF v1.x which is not actively supported and please try to upgrade to TF v2.4 or later.Thanks!
